### DISCUSSION TITLE
Spoolman collaboration

### DISCUSSION THREAD

**Discussion by Donkie** [markhaehnel/bambulab#15] [Ideas]
Hi, I'm the author of the Spoolman database/filament tracking service https://github.com/Donkie/Spoolman that has been gaining some traction lately. The main usage is to automatically track filament usage via klipper/moonraker.

A common request from the community though is to add support for Bambulab printers to track filament usage for their printers, which I understand are not controlled by a klipper/moonraker stack, but instead have their own software. I noticed though that you have managed to dig into their API, and it looks to me like this API would be able to do the same kind of tracking.

My idea is then to have a service like you have here that act as a bridge between the bambulab API and the Spoolman API, that people could spin up if they want to next to the existing systems.

Is this something that you would be interested in collaborating with me on? I personally don't have a bambulab printer so for me it's very hard to do something like this myself.

For reference, here is the Moonraker integration https://github.com/Arksine/moonraker/blob/master/moonraker/components/spoolman.py . Half of it is relating to a proxy that allows frontends to access Spoolman data more easily, the other is the filament tracking.

**Comment by markhaehnel** [comment]
Hi Donkie,
i would love to collaborate but i cannot say yet if it's possible to track filament usage somehow. The MQTT API of Bambulab is kinda limited. I will try to figure out if it's possible in the new year and will come back to you!

**Comment by d3nn3s08** [comment]
hallo also es werden sehr viele daten gesendet 

`{
  "name": "Layer Num",
  "icon": "mdi:layers",
  "device": {
    "identifiers": [
      "X1C_Bambi"
    ],
    "manufacturer": "Bambu Labs",
    "model": "X1C",
    "name": "X1C_Bambi"
  },
  "unique_id": "X1C_Bambi_layer_num",
  "object_id": "X1C_Bambi_layer_num",
  "state_topic": "homeassistant/sensor/X1C_Bambi/layer_num/state",
  "json_attributes_topic": "homeassistant/sensor/X1C_Bambi/layer_num/attr",
  "availability_topic": "homeassistant/sensor/X1C_Bambi/status/state"
}
`
![Screenshot 2024-02-02 204940](https://github.com/markhaehnel/bambulab/assets/96798557/82488c52-6f4d-4210-b350-2dccd9604d72)


Update
in HA Werden noch mehr Daten Ausgelesen Sie bild

[
![Screenshot 2024-02-02 205517](https://github.com/markhaehnel/bambulab/assets/96798557/5522aff1-446e-4bf3-99bd-ea2ab86c4f5a)
](url)

**Comment by MPJ1988** [comment]
Hello, are there any updates on the project? I'm really interested. I would like to have a filament storage and organization software for my X1C. What would be great, of course, is an integration of NFC, so that you can also create these for other filament manufacturers, for example via an app, and that the AMS can then read them. Unfortunately, I can't program, but I would like to help somehow.

**Reply by markhaehnel** [reply]
I'm afraid that I don't have the time to deal with a Spoolman integration.

The integration itself is outside of my library's responsibility and should be implemented inside Spoolman. The API itself should deliver enough information to get it implemented there using this library.

**Reply by rouvenschandl** [reply]
> Hello, are there any updates on the project? I'm really interested. I would like to have a filament storage and organization software for my X1C. What would be great, of course, is an integration of NFC, so that you can also create these for other filament manufacturers, for example via an app, and that the AMS can then read them. Unfortunately, I can't program, but I would like to help somehow.

Unfortunately, integrating NFC for filament storage and organization for your X1C is extremely difficult and unlikely. The NFC tags used by Bambu are encrypted and, as far as I know, cannot be replicated. This means it's not possible to create custom NFC tags that the AMS can read.

**Comment by Zriel88** [comment]
I think the idea of this conversation was to know if it was possible to track the filament usage of an active print via this library (if it was exposed somehow via MQTT).
If so, then the spoolman dev could integrate it when possible.

**Reply by operinko** [reply]
It's technically possible (with some best-guess shenanigans for real-time usage data based on current layer and such).
From what I can see, Bambu Lab exposes the finished print weight (estimated printed material) and print length (in meters), but no intermediate data. Both of those values update *only* at the end of the print (regardless of how the print ended).

**Reply by darkorb** [reply]
The `remain` portion of the spool data. Relies on it being a Bambu spool though and then it's updated as the print progresses.

**Comment by rouvenschandl** [comment]
I'm considering buying a Bambulab printer as a second printer. If I do, I could try / help to integrate it with Spoolman for filament tracking.

**Comment by alexandrovic-code** [comment]
I am really looking forward fort this integration. Updating once at the end of the print would be enough for now (not "live" during printing). Let me know if I can help somehow (A1 mini and Homeassistant/MQTT available).
